pkg/uprobetracer/bpf/usdt_helper.bpf.c

```diff
@@ -0,0 +1,363 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2024 The Inspektor Gadget authors
+
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+
+#define USDT_ARG_VERSION 1
+
+enum USDT_ARG_TYPE {
+	/* Argument in register */
+	USDT_ARG_TYPE_REG,
+	/* Argument in memory */
+	USDT_ARG_TYPE_MEM,
+};
+
+enum USDT_ARG_LENGTH {
+	USDT_ARG_LENGTH_UINT8 = 0,
+	USDT_ARG_LENGTH_INT8,
+	USDT_ARG_LENGTH_UINT16,
+	USDT_ARG_LENGTH_INT16,
+	USDT_ARG_LENGTH_UINT32,
+	USDT_ARG_LENGTH_INT32,
+	USDT_ARG_LENGTH_UINT64,
+	USDT_ARG_LENGTH_INT64,
+};
+
+// using inline asm to avoid the "dereference of modified ctx ptr" error,
+// see https://mejedi.dev/posts/ebpf-dereference-of-modified-ctx-ptr-disallowed/
+#define RETURN_CTX_FIELD(field)                                               \
+	asm("%[res] = *(u64 *)(%[base] + %[offset])"                          \
+	    : [res] "=r"(buf)                                                 \
+	    : [base] "r"(ctx), [offset] "i"(offsetof(struct pt_regs, field)), \
+	      "m"(*ctx));                                                     \
+	return buf;
+
+#if defined(__TARGET_ARCH_arm64)
+
+enum USDT_ARG_REG {
+	USDT_ARG_REG_ZERO,
+	USDT_ARG_REG_SP,
+	USDT_ARG_REG_R8,
+	USDT_ARG_REG_R9,
+	USDT_ARG_REG_R10,
+	USDT_ARG_REG_R11,
+	USDT_ARG_REG_R12,
+	USDT_ARG_REG_R13,
+	USDT_ARG_REG_R14,
+	USDT_ARG_REG_R15,
+	USDT_ARG_REG_R0,
+	USDT_ARG_REG_R1,
+	USDT_ARG_REG_R2,
+	USDT_ARG_REG_R3,
+	USDT_ARG_REG_R4,
+	USDT_ARG_REG_R5,
+	USDT_ARG_REG_R6,
+	USDT_ARG_REG_R7,
+	USDT_ARG_REG_R16,
+	USDT_ARG_REG_R17,
+	USDT_ARG_REG_R18,
+	USDT_ARG_REG_R19,
+	USDT_ARG_REG_R20,
+	USDT_ARG_REG_R21,
+	USDT_ARG_REG_R22,
+	USDT_ARG_REG_R23,
+	USDT_ARG_REG_R24,
+	USDT_ARG_REG_R25,
+	USDT_ARG_REG_R26,
+	USDT_ARG_REG_R27,
+	USDT_ARG_REG_R28,
+	USDT_ARG_REG_R29,
+	USDT_ARG_REG_R30,
```

Comment:
The strange sorting comes from [1], because x64 and arm64 are sharing some registers with the same name, I put the common parts in the front.

But yes, I would better to write two separate parser for each arch, and do not mix them together anymore. This is indeed confusing.

[1] https://github.com/inspektor-gadget/inspektor-gadget/pull/2765/files#diff-36fffe66de6363e6a7be945e08795a95ef1fdb096ee7959f759c523d9be54f28R40

Reply:
I'm now using the same sorting as the one in vmlinux.h